### PR TITLE
Verify checksum for the downloaded genesis block - Closes #1187

### LINKED ISSUE #1187: Verify checksum for the downloaded genesis block
### Description

In the case of large genesis blocks (mainnet / testnet), Lisk Service is unable to fetch the block directly from the node and needs to download it from a remote server which is customisable. Integrity checks for such files should be performed prior to processing the data.

### Acceptance Criteria

When unable to fetch the genesis block from the node
- Download the genesis block from the remote server
- Corresponding checksum file is downloaded
- Verify the checksum is a match and log the result
- Exit the application if the downloaded file cannot be validated

### PR DESCRIPTION
**Pull request by priojeetpriyom**
### What was the problem?
This PR resolves #1187 

### How was it solved?
- [x] match checksum file with the genesis block content
- [x] refactor previous implemetation for loading genesisblock
- [x] retry if the checksum match fails
- [x] move utility files functions under utils folder

### How was it tested?
Local

Steps:
paste following code in blocksUtils.js and execute the file.
```
genesisBlockUrl = config.networks[0].genesisBlockUrl;
genesisBlockFilePath = `./data/${config.networks[0].name}/genesis_block.json`;

downloadAndValidateGenesisBlock();

```

### CODE REVIEW

File: services/blockchain-connector/shared/utils/fs.js [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+	return readWrapper(filePath, 'utf8');
+};
+
+const deleteFileRecursive = async (filePath) => {
+	if (await exists(filePath)) {
+		const rmWrapper = util.promisify(fs.rm);
+
+		await rmWrapper(filePath, { recursive: true, force: true });
+	}
+};
```
Comment (sameersubudhi) [inline services/blockchain-connector/shared/utils/fs.js:82]: Please rename the methods to `readFile` and `deleteFile` and re-use the implementations from `services/export/shared/helpers/file.js`.

File: services/blockchain-connector/shared/utils/fs.js [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+const verifyFileChecksum = async (filePath, checksumPath) => {
+	const expectedChecksum = (await readFileAsync(checksumPath)).split(' ')[0];
+
+	return verifyChecksum(filePath, expectedChecksum);
+};
```
Comment (sameersubudhi) [inline services/blockchain-connector/shared/utils/fs.js:88]: We can remove this method. The `verifyChecksum` could directly take both file paths as input.
If we still want to maintain this method, please move it to the `download.js` file instead.

File: services/blockchain-connector/shared/sdk/blocksUtils.js [hunk 3]
```diff
@@ -80,28 +76,46 @@ const loadConfig = async () => {
...
-	if (!(await exists(directoryPath))) await mkdir(directoryPath, { recursive: true });
+	const genesisFileName = genesisBlockUrl.substring(genesisBlockUrl.lastIndexOf('/') + 1);
+	const genesisFilePath = `${directoryPath}/${genesisFileName}`;
+	const checksumFilePath = `${genesisFilePath}.SHA256`;
 
-	logger.info(`Downloading genesis block to the filesystem from: ${genesisBlockUrl}`);
+	do {
+		/* eslint-disable no-await-in-loop */
+		// Delete all previous files including the containing directory
+		await deleteFileRecursive(directoryPath);
```
Comment (sameersubudhi) [inline services/blockchain-connector/shared/sdk/blocksUtils.js:89]: Please delete the contents only when the checksum validation fails.
```suggestion
```

File: services/blockchain-connector/shared/sdk/blocksUtils.js [hunk 4]
```diff
@@ -80,28 +76,46 @@ const loadConfig = async () => {
...
-	await downloadFile(genesisBlockUrlSHA256, directoryPath);
+		// Verify the integrity of the downloaded file, retry on failure
+		const isValidGenesisBlock = await verifyFileChecksum(genesisFilePath, checksumFilePath);
+
+		// Extract the file if necessary
+		if (genesisFilePath.endsWith('.tar.gz')) {
+			await extractTarBall(genesisFilePath, directoryPath);
+		}
+
+		if (isValidGenesisBlock) return true;
```
Comment (sameersubudhi) [inline services/blockchain-connector/shared/sdk/blocksUtils.js:106]:
```suggestion
		if (isValidGenesisBlock) return true;
		
		// Delete the directory recursively if the validation fails
		await deleteFileRecursive(directoryPath);
```

File: services/blockchain-connector/shared/sdk/blocksUtils.js [hunk 5]
```diff
@@ -80,28 +76,46 @@ const loadConfig = async () => {
...
+		/* eslint-enable no-await-in-loop */
+	} while (retries-- > 0);
+
+	logger.fatal(`Could not verfiy genesis block. genesisBlockUrl:${genesisBlockUrl}`);
+	process.exit();
```
Comment (sameersubudhi) [inline services/blockchain-connector/shared/sdk/blocksUtils.js:111]:
```suggestion
	logger.fatal(`Unable to verify the integrity of the downloaded genesis block from ${genesisBlockUrl}`);
	logger.fatal('Exiting the application');
	process.exit(1);
```

File: services/blockchain-connector/shared/sdk/blocksUtils.js [hunk 6]
```diff
@@ -80,28 +76,46 @@ const loadConfig = async () => {
 	}
 };
 
-const downloadGenesisBlock = async () => {
+// eslint-disable-next-line consistent-return
+const downloadAndValidateGenesisBlock = async (retries = 2) => {
```
Comment (sameersubudhi) [inline services/blockchain-connector/shared/sdk/blocksUtils.js:80]: Please consider wrapping up the entire logic inside of a `try..catch` block.

**Review by sameersubudhi — CHANGES_REQUESTED** [review]

File: services/blockchain-connector/shared/sdk/blocksUtils.js (hunk 5 above)
Comment (priojeetpriyom) [reply services/blockchain-connector/shared/sdk/blocksUtils.js]: Applied.

File: services/blockchain-connector/shared/sdk/blocksUtils.js (hunk 4 above)
Comment (priojeetpriyom) [reply services/blockchain-connector/shared/sdk/blocksUtils.js:103]: Applied

File: services/blockchain-connector/shared/sdk/blocksUtils.js (hunk 3 above)
Comment (priojeetpriyom) [reply services/blockchain-connector/shared/sdk/blocksUtils.js]: Applied

File: services/blockchain-connector/shared/utils/fs.js (hunk 2 above)
Comment (priojeetpriyom) [reply services/blockchain-connector/shared/utils/fs.js]: wanted to support both ways for future. will merge as not required for now 🙏

File: services/blockchain-connector/shared/utils/fs.js [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+const remove = async (filePath) => {
+	if (await exists(filePath)) {
+		const rmWrapper = util.promisify(fs.rm);
+
+		await rmWrapper(filePath, { recursive: true, force: true });
+	}
+};
+
+const verifyFileChecksum = async (filePath, checksumPath) => {
+	const expectedChecksum = (await read(checksumPath)).split(' ')[0];
```
Comment (nagdahimanshu) [inline services/blockchain-connector/shared/utils/fs.js:85]:
```suggestion
	const [expectedChecksum] = (await read(checksumPath)).split(' ');
```

File: services/blockchain-connector/shared/utils/fs.js [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+			reject(error);
+		});
+		fileStream.on('end', () => {
+			resolve(dataHash.digest());
+		});
+	});
+
+	const fileChecksum = fileHash.toString('hex');
+	if (fileChecksum !== expectedChecksum) {
+		logger.info(`File checksum: ${fileChecksum} mismatched with expected checksum: ${expectedChecksum}. filepath:${filePath}`);
```
[the remark this answers is not in the document]
Comment (priojeetpriyom) [reply services/blockchain-connector/shared/utils/fs.js]: Applied

File: services/blockchain-connector/shared/utils/fs.js (hunk 2 above)
Comment (priojeetpriyom) [reply services/blockchain-connector/shared/utils/fs.js]: Applied.

File: services/blockchain-connector/shared/utils/fs.js (hunk 7 above)
Comment (priojeetpriyom) [reply services/blockchain-connector/shared/utils/fs.js]: Applied

File: services/blockchain-connector/shared/sdk/blocksUtils.js (hunk 6 above)
Comment (priojeetpriyom) [reply services/blockchain-connector/shared/sdk/blocksUtils.js:80]: Applied

File: services/blockchain-connector/shared/utils/fs.js [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+		return false;
+	}
+
+	return true;
+};
```
[the remark this answers is not in the document]
Comment (priojeetpriyom) [reply services/blockchain-connector/shared/utils/fs.js]: Applied.

File: services/blockchain-connector/shared/utils/fs.js (hunk 1 above)
Comment (priojeetpriyom) [reply services/blockchain-connector/shared/utils/fs.js]: Applied

File: services/blockchain-connector/shared/utils/download.js [hunk 10]
```diff
@@ -100,9 +103,43 @@ const downloadFile = (url, dirPath) => new Promise((resolve, reject) => {
 	});
 });
 
+const verifyFileChecksum = async (filePath, checksumPath) => {
```
Comment (sameersubudhi) [inline services/blockchain-connector/shared/utils/download.js:106]: Please add an `info` log in `downloadFile` to state that a file is being downloaded for a specific URL. Kindly ensure that other corresponding methods have the same as well.

File: services/blockchain-connector/shared/utils/download.js [hunk 11]
```diff
@@ -100,9 +103,43 @@ const downloadFile = (url, dirPath) => new Promise((resolve, reject) => {
...
+	// validate existance of both files
+	if (!(await exists(filePath)) || !(await exists(checksumPath))) {
+		logger.info(`Checksum verification failed for empty file(s). filePath:${filePath}, checksumPath:${checksumPath}`);
```
Comment (sameersubudhi) [inline services/blockchain-connector/shared/utils/download.js:109]:
```suggestion
		logger.info(`Checksum verification failed. One of the following files does not exist:\nfilePath:${filePath}\nchecksumPath:${checksumPath}`);
```

File: services/blockchain-connector/shared/utils/fs.js [hunk 12]
```diff
@@ -40,7 +41,34 @@ const mkdir = async (directoryPath, options = { recursive: true, mode: '0o777' }
...
+		return resolve(data);
+	});
+});
+
+const rm = async (directoryPath, options = {}) => {
+	await fs.rm(
+		directoryPath,
+		options,
+		(err) => {
+			if (err) logger.error(`Error when removing directory: ${directoryPath}\n`, err.message);
```
Comment (sameersubudhi) [inline services/blockchain-connector/shared/utils/fs.js:59]:
```suggestion
			if (err) {
				logger.error(`Error when removing directory: ${directoryPath}\n`, err.message);
				return false;
			}
			return true;
```

File: services/blockchain-connector/shared/sdk/blocksUtils.js [hunk 13]
```diff
@@ -80,28 +76,52 @@ const loadConfig = async () => {
...
+
+			// Verify the integrity of the downloaded file, retry on failure
+			const isValidGenesisBlock = await verifyFileChecksum(genesisFilePath, checksumFilePath);
+
+			// Extract the file if necessary
+			if (genesisFilePath.endsWith('.tar.gz')) {
+				await extractTarBall(genesisFilePath, directoryPath);
+			}
+
+			if (isValidGenesisBlock) return true;
```
Comment (sameersubudhi) [inline services/blockchain-connector/shared/sdk/blocksUtils.js:104]: Please extract the tarball only when the integrity check is successful.
```suggestion
			if (isValidGenesisBlock) {
				// Extract the file if necessary
				if (genesisFilePath.endsWith('.tar.gz')) {
					await extractTarBall(genesisFilePath, directoryPath);
				}

				return true;
			}
```

**Review by sameersubudhi — CHANGES_REQUESTED** [review]

File: services/blockchain-connector/shared/utils/download.js (hunk 11 above)
Comment (priojeetpriyom) [reply services/blockchain-connector/shared/utils/download.js]: Applied

File: services/blockchain-connector/shared/utils/fs.js (hunk 12 above)
Comment (priojeetpriyom) [reply services/blockchain-connector/shared/utils/fs.js:59]: Applied

File: services/blockchain-connector/shared/sdk/blocksUtils.js (hunk 13 above)
Comment (priojeetpriyom) [reply services/blockchain-connector/shared/sdk/blocksUtils.js]: Applied

File: services/blockchain-connector/shared/utils/download.js (hunk 10 above)
Comment (priojeetpriyom) [reply services/blockchain-connector/shared/utils/download.js:111]: Applied

File: services/blockchain-connector/shared/utils/download.js [hunk 14]
```diff
@@ -18,18 +18,22 @@ const http = require('http');
...
 } = require('lisk-service-framework');
 
+const { read, exists } = require('./fs');
+
 const logger = Logger();
 
 const getHTTPProtocolByURL = (url) => url.startsWith('https') ? https : http;
 
 const downloadAndExtractTarball = (url, directoryPath) => new Promise((resolve, reject) => {
+	logger.info(`Downloading external file.\nurl:${url}\ndirectory:${directoryPath}`);
```
Comment (sameersubudhi) [inline services/blockchain-connector/shared/utils/download.js:36]:
```suggestion
	logger.info(`Downloading and extracting file from ${url} to ${directoryPath}`);
```

File: services/blockchain-connector/shared/utils/download.js [hunk 15]
```diff
@@ -48,6 +52,7 @@ const downloadAndExtractTarball = (url, directoryPath) => new Promise((resolve,
 });
 
 const downloadJSONFile = (fileUrl, filePath) => new Promise((resolve, reject) => {
+	logger.info(`Downloading external file.\nurl:${fileUrl}\nfilePath:${filePath}`);
```
Comment (sameersubudhi) [inline services/blockchain-connector/shared/utils/download.js:55]:
```suggestion
	logger.info(`Downloading JSON file from ${url} to ${directoryPath}`);
```

File: services/blockchain-connector/shared/utils/download.js [hunk 16]
```diff
@@ -60,6 +65,7 @@ const downloadJSONFile = (fileUrl, filePath) => new Promise((resolve, reject) =>
 });
 
 const downloadAndUnzipFile = (fileUrl, filePath) => new Promise((resolve, reject) => {
+	logger.info(`Downloading external file.\nurl:${fileUrl}\nfilePath:${filePath}`);
```
Comment (sameersubudhi) [inline services/blockchain-connector/shared/utils/download.js:68]:
```suggestion
	logger.info(`Downloading and extracting file from ${url} as ${filePath}`);
```

File: services/blockchain-connector/shared/utils/download.js [hunk 17]
```diff
@@ -80,6 +86,8 @@ const downloadAndUnzipFile = (fileUrl, filePath) => new Promise((resolve, reject
 });
 
 const downloadFile = (url, dirPath) => new Promise((resolve, reject) => {
+	logger.info(`Downloading external file.\nurl:${url}\ndirectory:${dirPath}`);
```
Comment (sameersubudhi) [inline services/blockchain-connector/shared/utils/download.js:89]:
```suggestion
	logger.info(`Downloading file from ${url} to ${dirPath}`);
```

File: services/blockchain-connector/shared/utils/fs.js [hunk 18]
```diff
@@ -40,7 +41,35 @@ const mkdir = async (directoryPath, options = { recursive: true, mode: '0o777' }
...
+			return !err;
```
Comment (sameersubudhi) [inline services/blockchain-connector/shared/utils/fs.js:60]:
```suggestion
			return !!err;
```

**Review by sameersubudhi — CHANGES_REQUESTED** [review]

**Review by sameersubudhi — APPROVED** [review]
LGTM 👍

**Review by nagdahimanshu — APPROVED** [review]
Looks good 👍

**Pull request merged**